### PR TITLE
ci: speed up CI builds

### PR DESCRIPTION
**Pull request by vitorhugods**
By upgrading the JDK version to 17, enabling the parallel GC and caching dependencies between builds.

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Our CI is sloooooow.

### Causes

- We do _not_ cache gradle dependencies between builds for all workflows.
- We're using JDK 11
- We're not using parallel garbage collector

#### Caching
The [setup-java](https://github.com/actions/setup-java#caching-packages-dependencies) action is able to cache dependencies between builds. And the cache is smart enough to consider the hash of all important gradle files, so if the hash changes, the cache won't be used.

We can easily enable it by adding `cache: 'gradle'` to all missing workflows.

We had issues in the past and disabled the cache, mostly because this smart hashing thing wasn't in place and we were using transient dependencies early in development.

Also, it's easier to clear caches now, just going to [Actions/Cache](https://github.com/wireapp/kalium/actions/caches).

#### JDK upgrade

Since Gradle 7.3, JDK 17 is supported. It's not something we had in the beginning of the project, but it's a quick win now.

Adopt JDK is not a thing anymore, so I just opted for Azul's Zulu. We could maybe benchmark it and see if there's a significant difference between Zulu and Temurin and maybe be more scientific about it, but not today :)

#### Bonus

[Parallel Garbage Collector](https://docs.oracle.com/javase/8/docs/technotes/guides/vm/gctuning/parallel.html) is now enabled.

According to [Bitrise's analysis](https://bitrise.io/blog/post/switch-to-jdk-17-parallel-gc-and-speed-up-your-android-builds-by-9-20), the combo of upgrading to JDK 17 + enabling the Parallel GC resulted in improvements from 9% to 20% in their benchmarks using three different projects.

### Testing

As this build changed the `gradle.properties` file, all tests should run.

### Notes

The JDK switch + dependency caching should only affect CI.

The parallel GC affects our local builds, but it's compatible with JDK 11 that some developers are using on their local machines.

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by vitorhugods** [comment]
I will wait for this current build to finish, and then trigger `update branch` to see if the cache is being used for all workflows and no dependencies are re-downloaded.

File: gradle.properties [hunk 1]
```diff
@@ -1,4 +1,4 @@
-org.gradle.jvmargs=-Xmx2g
+org.gradle.jvmargs=-Xmx2g -XX:+UseParallelGC
```
Comment (gongracr) [inline gradle.properties:1]: Should we replicate the same on our `local.properties` or would this one override the local one?

**Review by gongracr — APPROVED** [review]
LGTM 🍨 Hoping that this can save us many hours of development! 🤞

**Review by ohassine — APPROVED** [review]

**Comment by vitorhugods** [comment]
For a comparison:

Android's `build the samples` step was taking around 10 minutes, as seen in [this build](https://github.com/wireapp/kalium/actions/runs/3823037845/jobs/6503781519) from a few minutes ago.

Now, [it takes around 3min40s](https://github.com/wireapp/kalium/actions/runs/3823488683/jobs/6504714133), plus the added 50s to grab the build cache.

Most of the improvement is cache related, no doubt. It's a bit tricky to compare the JDK change, but in any case, it's a step away from deprecation and towards modern tools with long-term-support anyway.

File: gradle.properties (hunk 1 above)
Comment (vitorhugods) [reply gradle.properties:1]: I'm not sure if `org.gradle.jvmargs` inside `local.properties` actually does anything 🤔.

`local.properties` is a Android Gradle Plugin thing, shouldn't affect "non-android-specific" things, unless added by other parties.

Like we did with the GitHub token thingy. We explicitly read from that file. But Gradle doesn't AFAIK.

In any case, this `global.properties` change should affect all developers.

**Pull request merged**